ion_functions/qc/qartod_functions.py

```diff
@@ -0,0 +1,385 @@
+    :param arr: An array of observed data
+    :param low_reps: number of repetitions prior to data being flagged suspect.
+    :param high_reps: number of repetitions prior to being flagged bad.
+    :param eps: a floating point number indicated the bounds within which the absolute
+                difference of the previous and current measurement will be considered a repeated vlaue.
+    :param prev_qc: An array of any previous QC values which were applied.  The first element is assumed to correspond
+                    to the position of the first element of arr
+    """
+    if not eps:
+        raise ValueError("Must specify a tolerance value (`eps`).")
+    if any([not isinstance(d, int) for d in [low_reps, high_reps]]):
+        raise TypeError("Both low and high repetitions must be type int.")
+    flag_arr = np.ones_like(arr, dtype='uint8')
+    if low_reps >= high_reps:
+        raise ValueError("Low reps must be less than high reps.")
+    # if the flag array is empty, just return the empty array.
+    # In numpy >= 1.11.2, numpy.nditer will throw a ValueError if you try
+    # to pass it an empty array, so prevent this here.
+    if arr.size == 0:
+        return flag_arr
+    it = np.nditer(arr)
+    # Consider moving loop to C for efficiency.
+    for elem in it:
+        idx = it.iterindex
+        # Check if low repetitions threshold is hit.
+        cur_flag = QCFlags.GOOD_DATA
+        if idx >= low_reps:
+            is_suspect = np.all(np.abs(arr[idx - low_reps:idx] - elem) < eps)
+            if is_suspect:
+                cur_flag = QCFlags.SUSPECT
+            # Since high reps is strictly greater than low reps, check it.
+            if is_suspect and idx >= high_reps:
+                is_bad = np.all(np.abs(arr[idx - high_reps:idx - low_reps]
+                                       - elem) < eps)
+                if is_bad:
+                    cur_flag = QCFlags.BAD_DATA
+        flag_arr[idx] = cur_flag
+    if prev_qc is not None:
+        set_prev_qc(flag_arr, prev_qc)
+    return flag_arr
+
+
+@add_qartod_ident(16, 'Time Series Flat Line Test')
+def time_series_flat_line_check(arr, low_reps=3, high_reps=5, eps=None, prev_qc=None):
+    """
+    Check for invariate observations and can be applied to all bulk wave
+    parameters.
+    :param arr: An array of observed data
+    :param low_reps: number of repetitions prior to data being flagged suspect.
+    :param high_reps: number of repetitions prior to being flagged bad.
+    :param eps: a floating point number indicated the bounds within which the absolute
+                difference of the previous and current measurement will be considered a repeated vlaue.
+    :param prev_qc: An array of any previous QC values which were applied.  The first element is assumed to correspond
+                    to the position of the first element of arr
+    """
+    return flat_line_check(arr, low_reps=low_reps, high_reps=high_reps,
```

Comment:
`time_series_flat_line_check` defaults `eps` to `None`, and then calls `flat_line_check` which will throw a `ValueError` if `eps` is `None`. Should we default `eps` to a reasonable value to avoid this? If not, the function header should specify that `eps` is a required parameter.